compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp

```diff
@@ -31,163 +30,75 @@ namespace mlir::iree_compiler::AMDAIE {
-  auto isIgnoredUser = [&](Operation *user) {
-    return origTiledAndFusedOps.count(user) || isa<tensor::DimOp>(user);
-  };
-
-  // 1. Tile the consumer.
-  SmallVector<OpResult> yieldedValuesToOrigValues;
-  SmallVector<Operation *> tiledOps;
-  rewriter.setInsertionPointAfter(rootOp);
-  FailureOr<scf::SCFTilingResult> tilingResult = scf::tileUsingSCFForallOp(
-      rewriter, cast<TilingInterface>(rootOp), options);
-  if (failed(tilingResult)) {
-    return failure();
+  scf::SCFTileAndFuseOptions tileAndFuseOptions;
+  tileAndFuseOptions.setTilingOptions(options);
+  tileAndFuseOptions.setFusionControlFn(
+      [&](tensor::ExtractSliceOp sliceOp, OpResult originalProducer,
+          bool isDestinationOperand) -> std::tuple<bool, bool> {
+        bool fusableOp =
+            TypeSwitch<Operation *, bool>(originalProducer.getOwner())
+                // List ops that shouldnt be fused.
+                .Case<tensor::PackOp, tensor::PadOp, linalg::CopyOp,
+                      memref::CopyOp>([](Operation *) { return false; })
+                // Fuse all Linalg ops (can be generalized later)
+                .Default([&](Operation *op) {
+                  return op->getDialect() ==
+                         rewriter.getContext()
+                             ->getLoadedDialect<linalg::LinalgDialect>();
+                });
+        return {fusableOp, false};
+      });
+
+  if (useFusion) {
+    tileAndFuseOptions.setFusionControlFn(
+        [&](tensor::ExtractSliceOp sliceOp, OpResult originalProducer,
+            bool isDestinationOperand) -> std::tuple<bool, bool> {
+          bool fusableOp =
+              TypeSwitch<Operation *, bool>(originalProducer.getOwner())
+                  // List ops that shouldnt be fused.
+                  .Case<tensor::PackOp, tensor::PadOp, linalg::CopyOp,
+                        memref::CopyOp>([](Operation *) { return false; })
+                  // Fuse all Linalg ops (can be generalized later)
+                  .Default([&](Operation *op) {
+                    return op->getDialect() ==
+                           rewriter.getContext()
+                               ->getLoadedDialect<linalg::LinalgDialect>();
+                  });
+          return {fusableOp, false};
+        });
   }
-  auto forallLoop = cast<scf::ForallOp>(tilingResult->loops[0]);
-  yieldedValuesToOrigValues.append(rootOp->result_begin(),
-                                   rootOp->result_end());
-  // A map from untiled value to scf.forall shared_outs. The shared_outs is used
-  // for DPS init operand if they use the same init operands.
-  llvm::DenseMap<Value, Value> mapToSharedOuts;
-  if (auto dpsOp = dyn_cast<DestinationStyleOpInterface>(rootOp)) {
-    for (auto [init, sharedOuts] :
-         llvm::zip_equal(dpsOp.getDpsInits(), forallLoop.getRegionOutArgs())) {
-      mapToSharedOuts[init] = sharedOuts;
-    }
+  // If user of pass requests they dont want fusion we disable all fusion.
+  else {
```

Comment:
There might be a better check here. We just need to make sure that if we are using tiling a reduction dimension, then we don't fuse.